cpp/src/io/parquet/page_data.cu

```diff
@@ -927,21 +939,40 @@ static __device__ bool setupLocalPageInfo(page_state_s* const s,
   int chunk_idx;
 
   // Fetch page info
-  if (t == 0) s->page = *p;
+  if (!t) s->page = *p;
   __syncthreads();
 
   if (s->page.flags & PAGEINFO_FLAGS_DICTIONARY) { return false; }
   // Fetch column chunk info
   chunk_idx = s->page.chunk_idx;
-  if (t == 0) { s->col = chunks[chunk_idx]; }
+  if (!t) { s->col = chunks[chunk_idx]; }
 
-  // zero nested value and valid counts
+  // if we can use the decode cache, set it up now
+  auto const can_use_decode_cache = s->page.nesting_info_size <= max_cacheable_nesting_decode_info;
+  if (can_use_decode_cache) {
+    int d = 0;
+    while (d < s->page.nesting_info_size) {
+      if (d + t < s->page.nesting_info_size) {
+        // these values need to be copied over from global
+        s->nesting_decode_cache[d + t].max_def_level = s->page.nesting_decode[d + t].max_def_level;
+        s->nesting_decode_cache[d + t].page_start_value =
+          s->page.nesting_decode[d + t].page_start_value;
+        s->nesting_decode_cache[d + t].start_depth = s->page.nesting_decode[d + t].start_depth;
+        s->nesting_decode_cache[d + t].end_depth   = s->page.nesting_decode[d + t].end_depth;
+      }
```

Comment:
Neither `d` or `t` are const. Is it safe to assume the compiler will cache this result or should we add `auto const idx = d + t;` to help it?

Reply:
> Neither `d` or `t` are const. Is it safe to assume the compiler will cache this result or should we add `auto const idx = d + t;` to help it?

Sure.  See  https://github.com/rapidsai/cudf/pull/12577#discussion_r1085540818